src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItemCollection.cs

```diff
@@ -277,6 +291,50 @@ private IList<ContentItem> FindItemsImplementation(PatternSet definition, IEnume
             return itemsList;
         }
 
+        private string GetRelatedFileExtensionProperty(string assemblyPath, IEnumerable<Asset> assets)
+        {
+            //E.g. if path is "lib/net472/A.B.C.dll", the prefix will be "lib/net472/A.B.C."
+            string assemblyPrefix = assemblyPath.Substring(0, assemblyPath.LastIndexOf('.') + 1);
+
+            if (_assemblyRelatedExtensions.TryGetValue(assemblyPrefix, out string relatedProperty))
+            {
+                return relatedProperty;
+            }
+
+            List<string> relatedFileExtensionList = null;
+            foreach (Asset asset in assets)
+            {
+                if (asset.Path is not null &&
+                    Path.GetExtension(asset.Path) != string.Empty &&
+                    //Assembly properties are files with extensions ".dll", ".winmd", ".exe", see ManagedCodeConventions.
+                    !Path.GetExtension(asset.Path).Equals(".dll", StringComparison.OrdinalIgnoreCase) &&
+                    !Path.GetExtension(asset.Path).Equals(".exe", StringComparison.OrdinalIgnoreCase) &&
+                    !Path.GetExtension(asset.Path).Equals(".winmd", StringComparison.OrdinalIgnoreCase) &&
+                    !asset.Path.Equals(assemblyPath, StringComparison.OrdinalIgnoreCase) &&
+                    asset.Path.StartsWith(assemblyPrefix, StringComparison.OrdinalIgnoreCase))
+                {
+                    if (relatedFileExtensionList is null)
+                    {
+                        relatedFileExtensionList = new List<string>();
+                    }
+                    relatedFileExtensionList.Add(asset.Path.Substring(assemblyPrefix.Length - 1));
+                }
+            }
+
+            // If no related files found.
+            if (relatedFileExtensionList is null || !relatedFileExtensionList.Any())
```

Comment:
Shouldn't add a null into the dictionary so we don't have to recalculate it again?